### PR TITLE
NEW: Adding feature #1167 for ::$allowed_parents to function like ::$allowed_children

### PR DESCRIPTION
**Pull request by patricknelson**
An implementation for #1167 on `3`, per @dhensby request.

### CODE REVIEW

**Comment by tractorcow** [comment]
Would it be ok to have some tests with this please? :D

**Comment by patricknelson** [comment]
Sure, you're referring to some demo code right? Here's what I had setup for myself:

**mysite/code/GenericPage.php**
```php
<?php

class GenericPage extends Page {
	// noop.
}
```

**mysite/code/SpecificParent.php**
```php
<?php

class SpecificParent extends Page {
	private static $allowed_parents = [];
	private static $allowed_children = ['SpecificChild'];
}
```

**mysite/code/SpecificChild.php**
```php
<?php

class SpecificChild extends Page {
	private static $allowed_parents = ['SpecificParent'];
	private static $allowed_children = [];
}
```

Or were you referring to unit tests?

**Comment by dhensby** [comment]
We need unit test coverage and possibly even behat tests too

**Comment by patricknelson** [comment]
Ok. I might need some guidance on how to do that for something like this. 

1. Would I be dynamically mocking up classes that extend `Page` and then somehow set their private static fields fields in the unit test itself? I suppose via fixtures (looking at some existing code) but didn't find any documentation or examples from what I've seen so far.

2. You'd want the tests in `cms/tests/model/SiteTreeTest.php` correct?

Thanks.

**Comment by patricknelson** [comment]
I'm confused on how this is possible since it goes against intuition (maybe magic behind the scenes via `ReflectionClass`?) but since I know private statics are accessible externally via `->stat()` does that mean I'd just need to use the fixtures YML, setup my test to get an object from that fixture and then set the static fields via `->set_stat()` (I just found)? If so then I think I know what to do.

**Comment by dhensby** [comment]
the stat functions should generally be avoided and `Config` used instead.

You want to:

1. Create some classes that have the `private static` vars set as you need them - these are typically defined at the end of the test file
2. Add a test functions to fetch those pages and attempt to relate them as parent-child as expected and ensure it works (no error when writing).
3. Add a test to check that pages that shouldn't work don't and throw an error as expected
4. Add a test to check edge cases, like this works on ancestor classes and so on.

**Comment by patricknelson** [comment]
Good point @dhensby. I've revisited this and realized that it should obviously follow similar syntax and logic as the `::$allowed_children` array. So the functionality will be that if you specify a class in this array (without a `*` prefix) it will include that class and any of its children as being allowed as parents to the current class. However, if you do provide the `*` prefix (e.g. `*ThisClassOnly`) it will only allow just that class and none of its children. I'll update the code I've got now to match that and then setup test cases to cover:

1. Ensuring any page is allowed as a parent if no `::$allowed_parents` array is provided.
2. Ensuring that a specified parent page **will** have another page as a child if that candidate child *does* list either:
   * The parent page (explicitly with the `*` prefix), or
   * The parent page or one of it's children (with no prefix).
3. Ensuring that **no** page is allowed as a parent to a class if the `::$allowed_parents` array is provided and is empty. Practically, this means that the `->allowedChildren()` method will never include this class in its returned results.

**Comment by patricknelson** [comment]
Ok, I've setup the test `->testAllowedParents()` on `SiteTreeTest` and so far it checks out. Let me know if you can think of any other scenarios that I may have missed (hopefully Travis CI checks out as well).

**Comment by patricknelson** [comment]
Ping @dhensby or @tractorcow I don't want this to get stale :)

File: tests/model/SiteTreeTest.php [hunk 1]
```diff
@@ -858,6 +858,25 @@ public function testAllowedChildren() {
...
 	}
 
+	public function testAllowedParents() {
+		$page = new Page();
+		$parent1 = new SiteTreeTest_SpecificParent1();
+		$parent2 = new SiteTreeTest_SpecificParent2();
+
+		// Ensure that none of the "Specific" parents and children show up as allowed children on a generic Page object.
+		$pageChildren = $page->allowedChildren();
+		$this->assertEquals(false, in_array("SiteTreeTest_SpecificParent1", $pageChildren));
```
Comment (dhensby) [inline tests/model/SiteTreeTest.php]: use `assertContains("SiteTreeTest_SpecificParent1", $pageChildren)` please

**Comment by dhensby** [comment]
Thanks for the test coverage, that was my main concern, I think.

I'm really busy at the moment, but will try to check this out (literally) and give it a test... Though can't promise how soon that'll be

File: tests/model/SiteTreeTest.php (hunk 1 above)
Comment (patricknelson) [inline tests/model/SiteTreeTest.php]: Ok, I'll use `assertNotContains()` since I'm trying to ensure it's not in the list.

**Comment by patricknelson** [comment]
No problem, thanks for following up. I've tweaked the unit test to use the more aesthetic and practical
`$this->assertNotContains(..., ...)` method as well.

**Comment by dhensby** [comment]
hey @patricknelson please can you rebase this and ping me when that's done? then I'll take a look :)

**Comment by patricknelson** [comment]
Alright, @dhensby done. Sorry for the delay in basically everything else. I've been super busy and preparing for a long vaca starting Friday :D.

**Comment by dhensby** [comment]
OK - I've checked this out locally and I think it's working as expected... but I was a bit confused by the following behaviour.

When I had a page with `allowed_parents` it could be a root page. I personally think this should STOP the page being added as a root page; there is of course the `can_be_root` config option (default true) but I feel this is a bit too un-expected.

We could get around this by demanding that `can_be_root` is explicitly set to true on the class in question.

@patricknelson The other thing we really need here is docs and perhaps more tests to assert this case and how we handle it? Behat tests would really be great too...

**Comment by patricknelson** [comment]
Right, I understand the confusion. I think documentation would be called for here. The logic behind allowing it to still exist in the root even though `::$allowed_parents = []` is because technically the root is not a parent (since in this case literally `ParentID = 0`). Also, from a standpoint of separation of concerns, if I say "This page cannot have parents," then that's what I want, it doesn't mean I cannot save it in the root; just that I cannot have parents. 

Also note that this would *actually* create a conflict with `::$can_be_root` since if an empty array for `::$allowed_parents` means you cannot save it in the root, then it literally cannot exist anywhere, creating more work when all I wanted to do is prevent it from having parents. In this case, it makes more sense to me to explicitly say `::$can_be_root = false` and at that point, if I as a developer have an issue finding a way to save my page, it's easier to resolve (because then I've explicitly said it cannot be root as well as defining that it cannot have parents). 

Does that make sense?

**Comment by patricknelson** [comment]
@dhensby  Also, side note: I have no clue how to use Behat and no prior knowledge on how to setup docs so neither of those will be today for sure (I'd be learning on the job on those). I can work on the other tests though. I'll ensure that you can still add `::$allowed_parents = []` to the root and maybe an alternate test case.

File: tests/model/SiteTreeTest.php [hunk 2]
```diff
@@ -1226,3 +1260,28 @@ public function augmentValidURLSegment() {
 	}
 
 }
+
+class SiteTreeTest_SpecificParent1 extends Page implements TestOnly {
+	// Cannot have any parents at all.
+	private static $allowed_parents = array();
```
Comment (tractorcow) [inline tests/model/SiteTreeTest.php]: I think we allow and test `$allowed_parents = 'none'`. We added an explicit non-falsey 'none' value in order to overcome the behaviour inherit in config at not being good at distinguishing falsey values.

SiteTree adds a default `private static $allowed_children = array("SiteTree");` so perhaps we should be consistent and add the same for allowed_parents?

This would mean your code would always expect a value for allowed_parents, rather than treating empty values as an implicit 'None', and reduce the risk of getting the 'wrong' empty values for root-only pages. :)

**Comment by tractorcow** [comment]
Sorry to prolong something that should have been solved a while back. I wish I got back to you earlier.

**Comment by chillu** [comment]
Looks useful, although I'm a bit worried about complicating any logic which needs to be run for every tree node. How much data does this add to each `data-allowedchildren` HTML attribute?

I assume this works the same way than `allowed_children` in that it only applies to the direct parent in the tree? So you can still place a child into a structure where a grand parent node might not match `allowed_parents`.

Also, did you check that the tree drag/drop interactions still work as expected? From memory, I've used some jstree APIs to deny dropping nodes based on the `data-allowedchildren` values.

How about docs? I know `allowed_children` aren't well documented, but we shouldn't continue this trend with `allowed_parents`, right? ;) https://docs.silverstripe.org/en/3.1/developer_guides/customising_the_admin_interface/how_tos/customise_cms_tree/

**Comment by dhensby** [comment]
@patricknelson sorry - I think I wasn't entirely clear.

I'm talking about when a page DOES have `allowed_parents` defined.

eg: `BlogPost` has `$allowed_parents = array('BlogHolder');` but blog page can still be at the root. Does allowed parents mean "these or root" or does it mean "can only be under these pages".

I think there are some scenarios we need to clear up:

1. Page has no allowed_parents defined (empty array/falsey value)
2. Page has allowed_parents set to "none" (proposal by @tractorcow)
3. Page has allowed_parents defined to one or more classes

For each scenario can the page be a root page and which pages can they be placed under?

I'd *personally* expect that if I define and "allowed_parent" then it means it has to be under one of those (ie: root page is not allowed).

I agree that if the allowed_parents is false then it'd be implicitly allowed at root level but that meaning it MUST be root, is different.

Personally, I'd prefer to add a 'must_be_root` option over implicitly forcing it through "none" value for `allowed_parents`

**Comment by tractorcow** [comment]
>  Personally, I'd prefer to add a 'must_be_rootoption over implicitly forcing it through "none" value forallowed_parents`

Well, for child pages, `none` means `must_be_leaf`, so it's just being consistent. :)

**Comment by dhensby** [comment]
Sure, I'm just saying I don't like this "none" value having a special meaning. I'd prefer that a config always has to be a single datatype

**Comment by tractorcow** [comment]
Then what should be done with `allowed_children` to make it consistent? I'd still rather have an API that's consistent with the rest of the API. Too many special rules will make it hard to use. `None can be used with allowed_children, but for allowed_parents you need to set an empty array` is arbitrary.

**Comment by dhensby** [comment]
> I suggest supporting both for both parameters, and in 4.x we can look at another solution.

agreed

**Comment by JorisDebonnet** [comment]
Looks interesting! I actually just needed this :)

Does this PR need more work, or is it perhaps rebased somewhere else?

**Comment by dhensby** [comment]
Hmm, we haven't come to a conclusion about inferred "can_be_root" or not...

I think keep this simple - if `can_be_root` is true, then it can be root even with defined `allowed_parents`.

Empty array or falsey value for `allowed_parents` means "can go anywhere"

`allowed_parents = "none"` means can't be a child page.

This work doesn't need to go against a different branch.

**Comment by chillu** [comment]
@patricknelson ping

**Comment by patricknelson** [comment]
Sorry about this -- I'll have to follow up on this in the coming weeks (not days); got a lot going on at the moment. Just letting you know so you'll be warned.

To be sure -- we're going to want to go with the implementation outlined by @dhensby here, correct? https://github.com/silverstripe/silverstripe-cms/pull/1171#issuecomment-148397150 If so I can move forward I think without any further instruction.

**Comment by dhensby** [comment]
> To be sure -- we're going to want to go with the implementation outlined by @dhensby here, correct? https://github.com/silverstripe/silverstripe-cms/pull/1171#issuecomment-148397150 If so I can move forward I think without any further instruction.

@chillu @tractorcow what do you think?

**Comment by tractorcow** [comment]
I like @dhensby 's solution yeah. ;D

Just be careful of falsey array values. I think allowed_children has a default value (defined in SiteTree) so it doesn't normally ever have the "is falsey check". It's ok if the apis differ a little there. ;)

I'm happy because we have the explicit "none" I was hoping for :D

**Comment by tractorcow** [comment]
@patricknelson any update on this feature? I believe we were looking at 'none' option for 'allowed_parents'?

**Comment by neilcreagh** [comment]
Bump - looks like this got so close!

**Comment by JorisDebonnet** [comment]
I guess the question is whether @patricknelson is willing to finish this with the final root/none/... remarks. If not, perhaps someone else needs to make that final commit.

**Comment by tractorcow** [comment]
If / when he comes back I'll re-open. ;P

**Comment by patricknelson** [comment]
Could you re-open this, please? I believe it made mention of merge conflicts which is not visible now (cannot recall 100%).

**Comment by patricknelson** [comment]
Oh ok I see why this has sat for so long. Some **TODO's** just for my reference (since there's such a huge backlog of comments and has sat for 3yrs):

### TODO:

- Fix merge conflicts
- Move away from `->stat()`
- Migrate to `$allowed_parents = 'none'` to follow current `$allowed_children = 'none'` syntax
- Possibly additional `->write()` tests to reiterate pages that shouldn't work still do not (already tested in retrieval in `->allowedChildren()`, but per @dhensby's [comment here](https://github.com/silverstripe/silverstripe-cms/pull/1171#issuecomment-76526993), maybe was missed?
- Verify and also initialize a `null` (or also empty, i.e. empty array?) definition of  `$allowed_parents` with `array("SiteTree")` if it's not defined


@dhensby @tractorcow @chillu  Please let me know if you happen to remember anything I've missed and I'll edit the list above.

---

**p.s.**

**Wrapping up on `$can_be_root`:**

@dhensby At this point I'm also reaffirming that no effort re: `$can_be_root` will be performed, in terms of code in this PR (per bullets above). I understand the confusion to a degree, but I know logically these two things (`$can_be_root` vs `$allowed_parents`) are unrelated in terms of implementation. Logically they come into conflict in a particular edge case scenario (i.e. `$can_be_root = false` while `$allowed_parents = "none"`), that requires no effort here. Just that we're aware that it could happen, particularly in cases of inheritance of these configs. 

> I think keep this simple - if can_be_root is true, then it can be root even with defined allowed_parents.

As I think this doesn't impact `->allowedChildren()` I presume I do nothing 😄

**Documentation:**

The best documentation I could find (that was documentation, not just a tutorial) was https://docs.silverstripe.org/en/3.0/reference/sitetree/ - What about v4? The search results [returned this hit ](https://docs.silverstripe.org/en/4/developer_guides/customising_the_admin_interface/how_tos/customise_cms_pages_list/) as the top hit for `allowed_children` for v4 and I don't find that as informative as the prior link.

Also re: documentation, @JorisDebonnet check out the weird hack from the link above:

```php
class Page extends SiteTree {
  // Don't let BlogEntry pages be underneath Pages.  Only underneath Blog holders.
  static $allowed_children = array("*Page,", "BlogHolder");
}
```

Coincidentally I think that would have been an ugly (but probably functional) workaround for a lack of `$allowed_parents` in the interim. Have you been using that? 😀

**Comment by JorisDebonnet** [comment]
> Check out the weird hack from the link above: `static $allowed_children = array("*Page,", "BlogHolder");` 
> Have you been using that? 😀

I have used that to e.g. create `Blog` and `BlogPost` extensions named `ActivityOverview` and `Activity`, and make sure that 'regular' `Blog`s can only have plain `BlogPost`s, while `ActivityOverview`s can only have `Activity`s. That was the most important bit to me to make sure I have pretty green buttons "Add BlogPost" and "Add Activity" on top of their GridFields, instead of a dropdown. 

Wasn't obvious though because the yaml config didn't seem to work and I had to use the following to restrict Blogs to only have plain BlogPosts:
```php
Config::inst()->remove('Blog', 'allowed_children');
Config::inst()->update('Blog', 'allowed_children', ['*BlogPost']);
```

But I could still create `BlogPost`s and `Activity`s in various other undesired places though. So I have not yet used it on the `Page` class itself, but I figure that would indeed work fine if overwritten where needed. It would potentially require a long list in `Page` though, if you have many PageTypes, since they'd all need to be listed except for those with an "allowed_parent". Something like `class Page { static $allowed_children = ["Gallery", "Blog", "ActivityOverview", "ContactFormPage","*Page"]; }`. I guess that's why you'd call it an ugly workaround :) 

Anyway, I have only used this on SilverStripe 3.x, not in SS4 yet.

**Comment by maxime-rainville** [comment]
Unfortunately we're only doing security updates to SilverStripe 3 from now on, so I'll have to close this one.

If you feel this is still a worthwhile feature, please refactor it so it works with SilverStripe 4.
